File: Solutions/[06]_2.java
import java.util.Scanner;

public class Main {

	public static int N, X, Y, result = 0;
	
	public static void solve(int n, int x, int y) {
		if (n == 2) {
			if (x == X && y == Y) {
				System.out.println(result);
				return;
			}
			result++;
			if (x == X && y + 1 == Y) {
				System.out.println(result);
				return;
			}
			result++;
			if (x + 1 == X && y == Y) {
				System.out.println(result);
				return;
			}
			result++;
			if (x + 1 == X && y + 1 == Y) {
				System.out.println(result);
				return;
			}
			result++;
			return;
		}
		solve(n / 2, x, y);
		solve(n / 2, x, y + n / 2);
		solve(n / 2, x + n / 2, y);
		solve(n / 2, x + n / 2, y + n / 2);
	}
	
	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);
		
		N = sc.nextInt();
		X = sc.nextInt();
		Y = sc.nextInt();
		
		solve((int) Math.pow(2, N), 0, 0);
	}

}

File: Solutions/[06]_2.py
# 시간제한이 추가되서 유효한 코드로 변경이 필요합니다
def solve(n, x, y):
    global result
    if n == 1:
        if x == X and y == Y:
            print(result)
        result += 1
        return

    n //= 2
    for i, j in [(x, y), (x, y+n), (x+n, y), (x+n, y+n)]:
        # 가지치기 추가: 4분면 중 1곳만 탐색한다
        if (i <= X < i+n) and (j <= Y < j+n):
            solve(n, i, j)
        else:
            result += n * n


result = 0
N, X, Y = map(int, input().split())
solve(2 ** N, 0, 0)
